docs/plugin/interfaces/guidelines.md

```diff
@@ -0,0 +1,383 @@
+        /* @brief Called if the job has done its work */
+        virtual void Completed() = 0;
+    };
+    Virtual core::hresult Process(ICallback* sink) = 0;
+    Virtual core::hresult Abort() = 0;
+};
+```
+
+Recommendations {#Recommendations} 
+===============
+
+State complete interfaces {#StatetCompleteInterfaces}
+-------------------------
+If a method on the interface can change an implementation into a different state, make sure the interface also has methods to revert the entered state. A good example is an interface that can ```core::hresult Start() = 0``` the executions of a job. As the execution of the job has an expected lifespan which is greater that the calls lifespan, make sure there is a method on the interface as well to bring back the interface to the initial, not started state. E.g. an ```core::hresult::Abort() = 0``` method. 
+
+Return ```core::hresult``` on methods {#ReturnCorehresultOnMethods}
+---------------------------------
+As most interfaces are designed to work over process boundaries, it is recommended to cater for issues during the passing of the process boundary, e.g. the other process might have crashed. This means that method calls on interfaces that seem triial but do change the state of an implementation, like ```SetTime()```, may fail due to cross process communication. If the return value of a method is defined as core::hresult, the Thunder COMRPC framework will signal issues with the communication by returning a negative value. The whole positive range from 0-2147483647 (0-0x7FFFFFFF) is free for the implementer to return values.
+
+Specifically define the length of the enum to be used {#SpecificallyDefineTheLenthgOfTheEnumToBeUsed}
+-----------------------------------------------------
+To allow for the code generator to optimize for the communication frame, it is recommended to always define the length of an enum if it is used as a parameter. If the enum length is not defined, it is assumed to be 32 bits. If an enum only has less than 256 values, it means 3 bytes are wasted on the line.
+
+Be verbose even if there is functionally no need {#BeVerboseEvenIfThereIsFunctionallyNoNeed}
+------------------------------------------------
+
+Describe/define the interface as verbose as possible. This is the contract between two parties, it is better to clearly state what the intention of a parameter is than to leave the intent floating. This means that if a parameter is expected ***not*** to be changed by the implementer define it as const, even if from a C/C++ language point of view it does not add value.
+
+``` c++
+core::hresult Logging(const string& cat, bool& enabled /* @out */) const = 0;
+core::hresult Logging(const string& cat, const bool enabled) = 0;
+```
+
+Although the const bool in the second method has no added value, it is a recommendation to use it, just to indicate clearly that the intend is that the implementation should not and can not change the parameter.
+
+!!! note "TODO"
```

Comment:
I see there is a Todo for properties. Just adding that we should provide guidelines on when to use properties vs specific getter/setter. I see some info under @property tag. Maybe include that in the guidelines here?